File: packages/flutter/android/src/main/kotlin/io/orch8/flutter/Orch8FlutterPlugin.kt
package io.orch8.flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.orch8.mobile.*

class Orch8FlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private var channel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private var engine: MobileEngine? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "io.orch8/mobile").also {
            it.setMethodCallHandler(this)
        }
        eventChannel = EventChannel(binding.binaryMessenger, "io.orch8/events").also {
            it.setStreamHandler(this)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel = null
        eventChannel?.setStreamHandler(null)
        eventChannel = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "resume" -> { engine?.resume(); result.success(null) }
            "pause" -> { engine?.pause(); result.success(null) }
            "start" -> handleStart(call, result)
            "sync" -> handleSync(call, result)
            "shutdown" -> { engine?.shutdown(); engine = null; result.success(null) }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun handleInitialize(call: MethodCall, result: MethodChannel.Result) {
        try {
            val args = call.arguments as Map<*, *>
            val dbPath = args["dbPath"] as? String ?: "orch8.db"

            val cfg = MobileEngineConfig(
                tickIntervalMs = (args["tickIntervalMs"] as? Number)?.toLong()?.toULong() ?: 100u,
                maxConcurrentSteps = (args["maxConcurrentSteps"] as? Number)?.toInt()?.toUInt() ?: 4u,
                maxStepsPerInstance = (args["maxStepsPerInstance"] as? Number)?.toInt()?.toUInt() ?: 1000u,
                maxConcurrentInstances = (args["maxConcurrentInstances"] as? Number)?.toInt()?.toUInt() ?: 10u,
                maxTickDurationMs = (args["maxTickDurationMs"] as? Number)?.toLong()?.toULong() ?: 5000u,
                maxInstanceLifetimeSecs = (args["maxInstanceLifetimeSecs"] as? Number)?.toLong()?.toULong() ?: 86400u,
                maxStoredSequences = (args["maxStoredSequences"] as? Number)?.toInt()?.toUInt() ?: 50u,
                maxSequenceSizeBytes = (args["maxSequenceSizeBytes"] as? Number)?.toLong()?.toULong() ?: 1_048_576u,
                handlerTimeoutMs = (args["handlerTimeoutMs"] as? Number)?.toLong()?.toULong() ?: 30000u,
                operationTimeoutMs = (args["operationTimeoutMs"] as? Number)?.toLong()?.toULong() ?: 10000u,
                telemetryEnabled = args["telemetryEnabled"] as? Boolean ?: true,
                environment = args["environment"] as? String ?: "production",
                rootPublicKey = args["rootPublicKey"] as? String ?: "",
                sdkVersion = args["sdkVersion"] as? String ?: "0.1.0",
            )

            engine = MobileEngine(dbPath, cfg)

            engine?.setListener(object : EngineListener {
                override fun onInstanceCompleted(instanceId: String, output: String) {
                    eventSink?.success(mapOf(
                        "type" to "instanceCompleted",
                        "instanceId" to instanceId,
                        "output" to output,
                    ))
                }

                override fun onInstanceFailed(instanceId: String, error: String) {
                    eventSink?.success(mapOf(
                        "type" to "instanceFailed",
                        "instanceId" to instanceId,
                        "error" to error,
                    ))
                }

                override fun onStepPending(instanceId: String, stepName: String, handler: String) {
                    eventSink?.success(mapOf(
                        "type" to "stepPending",
                        "instanceId" to instanceId,
                        "stepName" to stepName,
                        "handler" to handler,
                    ))
                }
            })

            result.success(null)
        } catch (e: Exception) {
            result.error("INIT_ERROR", e.message, null)
        }
    }

    private fun handleStart(call: MethodCall, result: MethodChannel.Result) {
        try {
            val args = call.arguments as Map<*, *>
            val instanceId = engine?.start(
                args["sequenceName"] as String,
                args["input"] as? String ?: "{}",
                args["dedupKey"] as? String,
            )
            result.success(instanceId)
        } catch (e: Exception) {
            result.error("START_ERROR", e.message, null)
        }
    }

    private fun handleSync(call: MethodCall, result: MethodChannel.Result) {
        try {
            val args = call.arguments as Map<*, *>
            val syncResult = engine?.sync(args["manifestUrl"] as String, null)
            result.success(mapOf(
                "added" to (syncResult?.added ?: 0),
                "updated" to (syncResult?.updated ?: 0),
                "removed" to (syncResult?.removed ?: 0),
                "skipped" to (syncResult?.skipped ?: 0),
            ))
        } catch (e: Exception) {
            result.error("SYNC_ERROR", e.message, null)
        }
    }
}

File: packages/flutter/ios/Classes/Orch8FlutterPlugin.swift
import Flutter
import UIKit
import Orch8Mobile

public class Orch8FlutterPlugin: NSObject, FlutterPlugin {
    private var engine: MobileEngine?
    private var eventSink: FlutterEventSink?
    private var channel: FlutterMethodChannel?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "io.orch8/mobile", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "io.orch8/events", binaryMessenger: registrar.messenger())

        let instance = Orch8FlutterPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(call.arguments as? [String: Any] ?? [:], result: result)
        case "resume":
            engine?.resume()
            result(nil)
        case "pause":
            engine?.pause()
            result(nil)
        case "start":
            handleStart(call.arguments as? [String: Any] ?? [:], result: result)
        case "sync":
            handleSync(call.arguments as? [String: Any] ?? [:], result: result)
        case "shutdown":
            engine?.shutdown()
            engine = nil
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleInitialize(_ args: [String: Any], result: FlutterResult) {
        let dbPath = args["dbPath"] as? String
            ?? FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
                .appendingPathComponent("orch8.db").path

        let config = MobileEngineConfig(
            tickIntervalMs: (args["tickIntervalMs"] as? NSNumber)?.uint64Value ?? 100,
            maxConcurrentSteps: (args["maxConcurrentSteps"] as? NSNumber)?.uint32Value ?? 4,
            maxStepsPerInstance: (args["maxStepsPerInstance"] as? NSNumber)?.uint32Value ?? 1000,
            maxConcurrentInstances: (args["maxConcurrentInstances"] as? NSNumber)?.uint32Value ?? 10,
            maxTickDurationMs: (args["maxTickDurationMs"] as? NSNumber)?.uint64Value ?? 5000,
            maxInstanceLifetimeSecs: (args["maxInstanceLifetimeSecs"] as? NSNumber)?.uint64Value ?? 86400,
            maxStoredSequences: (args["maxStoredSequences"] as? NSNumber)?.uint32Value ?? 50,
            maxSequenceSizeBytes: (args["maxSequenceSizeBytes"] as? NSNumber)?.uint64Value ?? 1_048_576,
            handlerTimeoutMs: (args["handlerTimeoutMs"] as? NSNumber)?.uint64Value ?? 30000,
            operationTimeoutMs: (args["operationTimeoutMs"] as? NSNumber)?.uint64Value ?? 10000,
            telemetryEnabled: args["telemetryEnabled"] as? Bool ?? true,
            environment: args["environment"] as? String ?? "production",
            rootPublicKey: args["rootPublicKey"] as? String ?? "",
            sdkVersion: args["sdkVersion"] as? String ?? "0.1.0"
        )

        do {
            engine = try MobileEngine(dbPath: dbPath, config: config)

            engine?.setListener(listener: FlutterEngineListener(plugin: self))

            result(nil)
        } catch {
            result(FlutterError(code: "INIT_ERROR", message: error.localizedDescription, details: nil))
        }
    }

    private func handleStart(_ args: [String: Any], result: FlutterResult) {
        do {
            let instanceId = try engine?.start(
                sequenceName: args["sequenceName"] as? String ?? "",
                input: args["input"] as? String ?? "{}",
                dedupKey: args["dedupKey"] as? String
            )
            result(instanceId)
        } catch {
            result(FlutterError(code: "START_ERROR", message: error.localizedDescription, details: nil))
        }
    }

    private func handleSync(_ args: [String: Any], result: FlutterResult) {
        do {
            let syncResult = try engine?.sync(
                manifestUrl: args["manifestUrl"] as? String ?? "",
                tokenProvider: nil
            )
            result([
                "added": syncResult?.added ?? 0,
                "updated": syncResult?.updated ?? 0,
                "removed": syncResult?.removed ?? 0,
                "skipped": syncResult?.skipped ?? 0,
            ])
        } catch {
            result(FlutterError(code: "SYNC_ERROR", message: error.localizedDescription, details: nil))
        }
    }

    fileprivate func sendEvent(_ event: [String: Any]) {
        eventSink?(event)
    }
}

extension Orch8FlutterPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = eventSink
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

private class FlutterEngineListener: EngineListener {
    private weak var plugin: Orch8FlutterPlugin?

    init(plugin: Orch8FlutterPlugin) {
        self.plugin = plugin
    }

    func onInstanceCompleted(instanceId: String, output: String) {
        plugin?.sendEvent([
            "type": "instanceCompleted",
            "instanceId": instanceId,
            "output": output,
        ])
    }

    func onInstanceFailed(instanceId: String, error: String) {
        plugin?.sendEvent([
            "type": "instanceFailed",
            "instanceId": instanceId,
            "error": error,
        ])
    }

    func onStepPending(instanceId: String, stepName: String, handler: String) {
        plugin?.sendEvent([
            "type": "stepPending",
            "instanceId": instanceId,
            "stepName": stepName,
            "handler": handler,
        ])
    }
}
